### PR TITLE
[`flake8-bugbear`] Implement `return-in-generator` (`B901`)

### PR DESCRIPTION
**Pull request by tobb10001**
## Summary

This PR implements the rule B901, which is part of the opinionated rules of `flake8-bugbear`.

This rule seems to be desired in `ruff` as per https://github.com/astral-sh/ruff/issues/3758 and https://github.com/astral-sh/ruff/issues/2954#issuecomment-1441162976.

## Test Plan

As this PR was made closely following the [CONTRIBUTING.md](https://github.com/astral-sh/ruff/blob/8a25531a7144fd4a6b62c54efde1ef28e2dc18c4/CONTRIBUTING.md), it tests using the snapshot approach, that is described there.

## Sources

The implementation is inspired by [the original implementation in the `flake8-bugbear` repository](https://github.com/PyCQA/flake8-bugbear/blob/d1aec4cbef7c4a49147c428b7e4a97e497b5d163/bugbear.py#L1092). The error message and [test file](https://github.com/PyCQA/flake8-bugbear/blob/d1aec4cbef7c4a49147c428b7e4a97e497b5d163/tests/b901.py) where also copied from there.

The documentation I came up with on my own and needs improvement. Maybe the example given in https://github.com/astral-sh/ruff/issues/2954#issuecomment-1441162976 could be used, but maybe they are too complex, I'm not sure.

## Open Questions

- [ ] Documentation. (See above.)

- [x] Can I access the parent in a visitor?

    The [original implementation](https://github.com/PyCQA/flake8-bugbear/blob/d1aec4cbef7c4a49147c428b7e4a97e497b5d163/bugbear.py#L1100) references the `yield` statement's parent to check if it is an expression statement. I didn't find a way to do this in `ruff` and used the `is_expresssion_statement` field on the visitor instead. What are your thoughts on this? Is it possible and / or desired to access the parent node here?

- [x] Is `Option::is_some(...)` -> `...unwrap()` the right thing to do?

    Referring to [this piece of code](https://github.com/tobb10001/ruff/blob/9d5a280f71103ef33df5676d00a6c68c601261ac/crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs?plain=1#L91-L96). From my understanding, the `.unwrap()` is safe, because it is checked that `return_` is not `None`. However, I feel like I missed a more elegant solution that does both in one.

## Other

I don't know a lot about this rule, I just implemented it because I found it in a https://github.com/astral-sh/ruff/labels/good%20first%20issue.

I'm new to Rust, so any constructive critisism is appreciated.

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                // do not recurse into nested functions, as they are evaluated
+                // individually
+            }
+            Stmt::Return(ast::StmtReturn { value, range }) => {
+                if Option::is_some(value) {
+                    self.return_ = Some(*range);
+                }
+            }
+            _ => {
+                self.in_expr_statement = stmt.is_expr_stmt();
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs:63]: I think I would do this by checking `Stmt::Expr` here, and then just introspecting the value directly to see if the value is `Expr::Yield`.

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/// B901
+pub(crate) fn return_x_in_generator(checker: &mut Checker, function_def: &StmtFunctionDef) {
+    if function_def.name.id == "__await__" {
+        return;
+    }
+
+    let mut visitor = ReturnXInGeneratorVisitor::default();
+    visitor.visit_body(&function_def.body);
+
+    if Option::is_some(&visitor.return_) && visitor.has_yield {
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs:91]: (Separately, `visitor.return_.is_some()` would be more natural than `Option::is_some` -- we can call it as a method.)

**Comment by charliermarsh** [comment]
Thanks! Two small comments based on your questions.

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tobb10001) [reply crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs]: Thanks for the hint.

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs (hunk 1 above)
Comment (tobb10001) [reply crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs:63]: Good idea, I updated the implementation to this.

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_bugbear/rules/return_x_in_generator.rs]: Oh yeah, if you don't need it def omit -- sometimes you need it, was hard to tell when writing on GitHub :)

**Review by charliermarsh — APPROVED** [review]
Thanks!

**Pull request merged**